### PR TITLE
Added a bugzilla attach python script

### PR DESCRIPTION
**Pull request by alevitter**
This script uploads the result of the must-gather command as an archived attachment to a supplied Bugzilla bug. Now has been tested against Red Hat Bugzilla.
The script trims any logfile down to the most recent lines based on their timestamp (using the -t parameter, default is 30 minutes) in order to reduce the size of the attachment (the maximum size is 19.5MB), but the untrimmed versions will remain in the output directory.
You can pass an API key as an optional parameter, or by setting BUGZILLA_API_KEY environment variable
```
usage: bugzilla_attach.py [-h] [--image image] [--api-key api-key] [-t TIME]
                          [--log-folder log-folder] [-r] [-i]
                          id

Sends the result of must-gather to Bugzilla.

positional arguments:
  id                    The ID of the bug in Bugzilla

optional arguments:
  -h, --help            show this help message and exit
  --image image         The image to use for must-gather
  --api-key api-key     Bugzilla API key. Can also be set using
                        BUGZILLA_API_KEY environment variable
  -t TIME, --time TIME  Number of minutes to use for trimming the log files.
                        Defaults to 30
  --log-folder log-folder
                        Optional destination for the must-gather output
                        (defaults to creating gather-files/ in the local
                        directory)
  -r, --reuse-must-gather
                        Use this to skip rerunning must-gather and just attach
                        what is already gathered
  -i, --interactive     Use this flag to prompt for a username and password.
                        Using this will prompt for retries if the login is
                        unsuccessful
```

Requires a working Openshift 4.1+ cluster (with any environment variables it needs to execute commands locally)
To generate an API key, go to your account > Preferences > API Keys. Check "Generate API Key with optional description" (and enter a description if you want) and then click "Submit Changes".

Solves https://github.com/kubevirt/must-gather/issues/29

### CODE REVIEW

File: scripts/bugzilla_attach.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        print("Creating archive: " + archive_name)
+        tar.add(LOGFOLDER, arcname=os.path.basename(LOGFOLDER))
+
+    print("Preparing to send the data to " + BUGZILLA_URL)
+
+    file_data = ""
+    with open(archive_name, "rb") as data_file:
+        file_data = base64.b64encode(data_file.read())
+
+    bugzilla_username = input("Enter Bugzilla username: ")
```
Comment (masayag) [inline scripts/bugzilla_attach.py]: that's not ideal if we'd like to use this script in automated env.
please expose required input from environment variables and as input parameters.

File: scripts/bugzilla_attach.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    # Make a new log folder
+    os.mkdir(LOGFOLDER)
+
+    # Open the output file
+    with open(LOGFOLDER + OUTPUT_FILE, "w+") as out_file:
+        # Run oc adm must-gather with the appropriate image and dest-dir
+        print("Running must-gather")
+        subprocess.run(
+            ["oc", "adm", "must-gather",
+            "--image=quay.io/kubevirt/must-gather", "--dest-dir=" + LOGFOLDER],
```
Comment (masayag) [inline scripts/bugzilla_attach.py]: --image / --images should be given as command line argument. the image might be for ocs / upstream vs downstream. therefore shouldn't be hard-coded in the script.

File: scripts/bugzilla_attach.py (hunk 1 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: In that case, it might be better to add an optional parameter that can take an API key (generated from the Bugzilla website), in which case the prompt for username/password would be skipped.

File: scripts/bugzilla_attach.py (hunk 2 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: --image is now an optional parameter which defaults to the kubevirt/must-gather image

File: scripts/bugzilla_attach.py (hunk 1 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Added an optional flag that allows passing an API key instead of prompting for a username and password

File: scripts/bugzilla_attach.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+    if args.image:
+        image = args.image
+    else:
+        image = IMAGE
+
+    use_api_key = args.api_key != None
+
+    # If the log folder already exists, delete it
+    if os.path.isdir(LOGFOLDER):
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: Would be good to parametrize this and use `LOGFOLDER` value as default.

File: scripts/bugzilla_attach.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        shutil.rmtree(LOGFOLDER)
+
+    # Make a new log folder
+    os.mkdir(LOGFOLDER)
+
+    # Open the output file
+    with open(LOGFOLDER + OUTPUT_FILE, "w+") as out_file:
+        # Run oc adm must-gather with the appropriate image and dest-dir
+        print("Running must-gather")
+        subprocess.run(
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: Please check the return code and print stderr if any. There is no need to continue if the command failed.

File: scripts/bugzilla_attach.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            print("Incorrect username or password.")
+            bugzilla_username = input("Username (leave blank to exit): ")
+            if bugzilla_username == "":
+                print("Username left blank, exiting")
+                exit()
+            bugzilla_password = getpass(prompt="Password: ")
+            resp = send_data(bugzilla_username, bugzilla_password, bug_id, file_data)
+            resp_json = resp.json()
+        # 101: Invalid bug id
+        elif resp_json["code"] == 101:
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: what about any other error code?

File: scripts/bugzilla_attach.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        "login": username,
+        "password": password,
+        "ids": [bug_id],
+        "summary": "Result from must-gather command",
+        "content_type": "application/gzip",
+        "data": file_data
+    }
+    return requests.post(url, json=data, headers=HEADERS)
+
+def send_data_with_api_key(api_key, bug_id, file_data):
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: Both functions `send_data` and `send_data_with_api_key` are very similar. Please unify.

File: scripts/bugzilla_attach.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        "ids": [bug_id],
+        "summary": "Result from must-gather command",
+        "content_type": "application/gzip",
+        "data": file_data
+    }
+    return requests.post(url, json=data, headers=HEADERS)
+
+def get_lines(file):
+    """Gets the number of lines in the file handle"""
+    return sum(1 for line in file)
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: Why no to use `len(file.readlines())`?

File: scripts/bugzilla_attach.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+import os
+import shutil
+import subprocess
+import tarfile
+import datetime
+import base64
+from getpass import getpass
+import requests
+
+MAX_LOGLINES = 100000
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: How did you come up with this number?

File: scripts/bugzilla_attach.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        "data": file_data
+    }
+    return requests.post(url, json=data, headers=HEADERS)
+
+def get_lines(file):
+    """Gets the number of lines in the file handle"""
+    return sum(1 for line in file)
+
+def trim_file(file, num_lines):
+    """Trims the file handle to the number of lines"""
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: What exactly do you want to do here?

File: scripts/bugzilla_attach.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+                        help="The ID of the bug in Bugzilla")
+    parser.add_argument("--image", metavar="image",
+                        help="The image to use for must-gather")
+    parser.add_argument("--api-key", metavar="api-key",
+                        help="Optional API key instead of username and login (will disable prompts to retry)")
+    args = parser.parse_args()
+
+    bug_id = args.ID
+
+    if args.image:
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: What do you think about accepting comma separated list of image names?

File: scripts/bugzilla_attach.py (hunk 3 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Done in recent commit

File: scripts/bugzilla_attach.py (hunk 4 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Done in recent commit

File: scripts/bugzilla_attach.py (hunk 5 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: These are the ones that as of now I can test, once I have access to somewhere I can test that attachment is working I can try to deal with other error codes.

Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: Please handle any other code in else block as generic issue.

File: scripts/bugzilla_attach.py (hunk 7 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Changed to this in recent commit

File: scripts/bugzilla_attach.py (hunk 8 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: I'm trying to keep the compressed size of the files down to around 4MB, and 100,000 lines gives around that (give or take based on the file)

File: scripts/bugzilla_attach.py (hunk 9 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Without trimming the files, the compressed archive is easily over 40MB, while the maximum file size for Bugzilla attachments is 19.5MB. So by taking only at most the last 100,000 lines of each file, I'm trying to reduce the size of the attachment while still preserving the most recent log lines and still going back far enough to be useful.

File: scripts/bugzilla_attach.py (hunk 10 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Do you mean running must-gather multiple times with different images? must-gather only accepts one image at a time. And given the size of the output from a single execution of must-gather, I don't think they would be able to be sent in the same attachment.

Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: 4.2 oc command provides [ability](https://github.com/openshift/oc/blob/master/pkg/cli/admin/mustgather/mustgather.go#L82) to use imageStream. We need to provide ability to use OCP, CNV or OCS images.

File: scripts/bugzilla_attach.py (hunk 9 above)
Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: I suggest to use [islice](https://docs.python.org/3/library/itertools.html#itertools.islice). It should simplify your code.

In general I would suggest to use file split instead. You do not know in which part of the file we can find valuable information. When you would use more images to collect data it would take more space so 100k lines maybe too much.

File: scripts/bugzilla_attach.py (hunk 5 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Done in recent commit; any other code will display the error that Bugzilla returns and will exit

File: scripts/bugzilla_attach.py (hunk 6 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: I'm not so sure unifying them is possible; they may look very similar but the data they are sending has different parameters.

Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: You could pass authentication details (api_key or username/password) as dictionary and merge with data dictionary.

File: scripts/bugzilla_attach.py (hunk 9 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: The issue with splitting the file is that I'd still need some way to figure out which part to upload. I can split the file and not add the first parts to the archive, but that still has the same issue of which is the important part.
As for whether 100k lines is too much, this is something that can be changed, or maybe it can be calculated based on the sizes of the logs generated.
Alternatively, the script can skip trimming altogether and split the archive into multiple attachments that are sent separately. But even on a simple cluster, an untrimmed archive has so far reached 66MB. I don't know if we want to be uploading potentially dozens of archives.

Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: My point is that we may trim important data. Usually valuable information is closed to the end of the log file not the beginning. Compressing and splitting seems to be much better than trimming the files.

**Comment by dankenigsberg** [comment]
I feel uncomfortable with this script, as it does too many things. Trimming must-gather content may be helpful; it may also drop the most interesting part of the logs. But I don't find it particularly urgent.

As for attaching a file to bugzilla, why do we need a new script when https://github.com/python-bugzilla/python-bugzilla own script is available on many client systems? I don't think we should "compete" with it. I'd rather just use it (if we have to) rather implementing a new command line using the library.

File: scripts/bugzilla_attach.py (hunk 10 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: When 4.2 is released, this can be updated to add that functionality

**Comment by alevitter** [comment]
> I feel uncomfortable with this script, as it does too many things. Trimming must-gather content may be helpful; it may also drop the most interesting part of the logs. But I don't find it particularly urgent.

The question of what to trim is still being decided, for now the decision is to do something that definitely reduce the size the attachment so that it can be uploaded successfully. This is also potentially a more general issue with must-gather: it generates a very verbose log that makes it harder to analyse.

> As for attaching a file to bugzilla, why do we need a new script when https://github.com/python-bugzilla/python-bugzilla own script is available on many client systems? I don't think we should "compete" with it. I'd rather just use it (if we have to) rather implementing a new command line using the library.

That script is using a deprecated API for accessing Bugzilla, as indicated [here](https://bugzilla.readthedocs.io/en/latest/integrating/apis.html#xml-rpc).

File: scripts/bugzilla_attach.py (hunk 9 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: The trimming so that the _last_ 100,000 lines are kept, not the _first_. Sorry if that was confusing. I effectively implemented `tail -n 100000 file > file`

**Comment by dankenigsberg** [comment]
> > I feel uncomfortable with this script, as it does too many things. Trimming must-gather content may be helpful; it may also drop the most interesting part of the logs. But I don't find it particularly urgent.
> 
> The question of what to trim is still being decided, for now the decision is to do something that definitely reduce the size the attachment so that it can be uploaded successfully. This is also potentially a more general issue with must-gather: it generates a very verbose log that makes it harder to analyse.

You second sentence adds to my discomfort. The trimming script should probably be propsed to mast-gather itself.

> 
> > As for attaching a file to bugzilla, why do we need a new script when https://github.com/python-bugzilla/python-bugzilla own script is available on many client systems? I don't think we should "compete" with it. I'd rather just use it (if we have to) rather implementing a new command line using the library.
> 
> That script is using a deprecated API for accessing Bugzilla, as indicated [here](https://bugzilla.readthedocs.io/en/latest/integrating/apis.html#xml-rpc).

This sounds like something you'd want to fix in /usr/bin/bugzilla, not a reason for a fork.

**Comment by pkliczewski** [comment]
> > > I feel uncomfortable with this script, as it does too many things. Trimming must-gather content may be helpful; it may also drop the most interesting part of the logs. But I don't find it particularly urgent.
> > 
> > 
> > The question of what to trim is still being decided, for now the decision is to do something that definitely reduce the size the attachment so that it can be uploaded successfully. This is also potentially a more general issue with must-gather: it generates a very verbose log that makes it harder to analyse.
> 
> You second sentence adds to my discomfort. The trimming script should probably be propsed to mast-gather itself.
> 
must-gather is here to collect everything that is needed. We are going to target objects with different tool.

File: scripts/bugzilla_attach.py (hunk 9 above)
Comment (brainfunked) [reply scripts/bugzilla_attach.py]: I agree with @pkliczewski. We should not be discarding any data without verifying that it is unneeded.

**Comment by alevitter** [comment]
> > > I feel uncomfortable with this script, as it does too many things. Trimming must-gather content may be helpful; it may also drop the most interesting part of the logs. But I don't find it particularly urgent.
> > 
> > 
> > The question of what to trim is still being decided, for now the decision is to do something that definitely reduce the size the attachment so that it can be uploaded successfully. This is also potentially a more general issue with must-gather: it generates a very verbose log that makes it harder to analyse.
> 
> You second sentence adds to my discomfort. The trimming script should probably be propsed to mast-gather itself.
> 
> > > As for attaching a file to bugzilla, why do we need a new script when https://github.com/python-bugzilla/python-bugzilla own script is available on many client systems? I don't think we should "compete" with it. I'd rather just use it (if we have to) rather implementing a new command line using the library.
> > 
> > 
> > That script is using a deprecated API for accessing Bugzilla, as indicated [here](https://bugzilla.readthedocs.io/en/latest/integrating/apis.html#xml-rpc).
> 
> This sounds like something you'd want to fix in /usr/bin/bugzilla, not a reason for a fork.

I agree that in theory they should be fixed in the proper places, but ultimately they presently are not. And if they cannot be fixed there, this script would need to operate the way it does.
When they are changed, the script can be changed as well.

File: scripts/bugzilla_attach.py (hunk 8 above)
Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: How difficult would it be to base it on time? We could parse the time in log and get only last 5 mins (configurable).

File: scripts/bugzilla_attach.py (hunk 10 above)
Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: OK, let's add this as separate PR

File: scripts/bugzilla_attach.py (hunk 5 above)
Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: :+1:

File: scripts/bugzilla_attach.py (hunk 9 above)
Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: What do you think about trimming based on time? In the worst case we would not be able to upload to bugzilla.

Comment (alevitter) [reply scripts/bugzilla_attach.py]: Not all files are formatted based on a timestamp. For the largest ones (the node logs), I've successfully been able to trim based on a given amount of time, but I'm not sure I can assume that those are going to be the only massive logs.

Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: Let's explore it. I see potential to reuse this code in other places.

**Comment by alevitter** [comment]
> I think that adding time based trimming would make this PR mergable. In next PR we would provide ability to provide multiple images.
The problem with trimming by time is that not all logs follow the same pattern. I would need to see how the output of must-gather looks in a few instances before being able to know which logs might be a problem (and then trim just those). As of now, only the node logs are a problem and I've written code that can trim those down but it relies on the date-time format used there. I don't know if there might be other large logs in other cases.

**Comment by pkliczewski** [comment]
Timestamp format it is something which requires investigation. We need this for other work that we have in the pipeline. Let's investigate and see which logs are problematic.

**Comment by masayag** [comment]
Let's examine if the format might vary based on locale or other configuration and see how trimming should be done and for which input: if requiring last 2 hours, it is more trivial to calculate than providing a date that might be in a different timezone than the one used by the cluster.

**Comment by alevitter** [comment]
Timezones are easier to deal with, since Python's datetime library can parse a timezone that's included in the timestamp. What's harder to deal with is what happens if there's no year information in the timestamp- this could make dealing with a log that spans a new year very hard to parse. Ideally I could simply find a timestamp in epoch time (not necessarily at the start of the line) but I've already seen logs that have neither a year nor an epoch time in most of their lines.

**Comment by pkliczewski** [comment]
@alevitter Please take a look [here](https://github.com/pkliczewski/fileops/blob/master/fileops/fileops.go#L131-L163) how I handled missing year and timezone for journal based logs. There are still qemu and cdi deployment logs which are problematic to parse. I simply ignore them.

File: scripts/bugzilla_attach.py [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+def check_bug_exists(bug_id):
+    """Checks whether the bug exists in Bugzilla"""
+    url = BUGZILLA_URL + '/rest/bug/%s' % bug_id
+    return "error" not in requests.get(url).json()
+
+
+def trim_file_by_time(filename, num_seconds, timestamp_regex, timestamp_format):
+    """Uses a binary search to locate where in the file to trim"""
+    with open(filename, "r+") as file:
+        lines = [line.rstrip('\n') for line in file]
```
Comment (pkliczewski) [inline scripts/bugzilla_attach.py]: Why not to reverse the lines and start from the end to the time we wanted?

Comment (alevitter) [reply scripts/bugzilla_attach.py]: The binary search is much faster than traversing from the end, since traversing from the end would need comparison of each line (which could be thousands) whereas even on large files I've had fewer than 25 comparisons.

Comment (pkliczewski) [reply scripts/bugzilla_attach.py]: OK

**Comment by pkliczewski** [comment]
@masayag The change looks good to me, is there anything else you want to change?

File: scripts/bugzilla_attach.py [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+This script attaches the result of the must-gather command, as executed
+by the kubevirt must-gather image, to the supplied bugzilla id on the
+Red Hat bugzilla website.
+It first creates an output subdirectory in the working directory named
+gather-files/ and then executes the following command:
+'oc adm must-gather --image=quay.io/kubevirt/must-gather
+--dest-dir=gather-files/' and pipes the output to
+gather-files/must-gather.log
+In order to meet the maximum attachment file sizes, logs are trimmed to the
+last n seconds (defaulting to 30 minuts)
```
Comment (masayag) [inline scripts/bugzilla_attach.py]: minuts --> minutes

File: scripts/bugzilla_attach.py [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+
+
+def trim_logs(logfolder, num_seconds):
+    for subdir, _, files in os.walk(logfolder):
+        for file in files:
+            if file == "must-gather.log": #Ignore the log made by capturing the output of must-gather
+                continue
+            if ".log" in file:
+                trim_file_by_time(os.path.join(subdir, file), num_seconds, PODLOG_TIMESTAMP_REGEX, PODLOG_TIMESTAMP_FORMAT)
+            if "kubelet" in file or "NetworkManager" in file:
```
Comment (masayag) [inline scripts/bugzilla_attach.py]: shouldn't there be 'else' here? or this condition should be checked first and its 'else' part of should be line 237 ?

File: scripts/bugzilla_attach.py [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+def try_parse_int(value):
+    """Tries to parse the value as an int"""
+    try:
+        return int(value), True
+    except ValueError:
+        return value, False
+
+
+
+def send_data(bug_id, file_name, file_data, comment, authentication):
```
Comment (masayag) [inline scripts/bugzilla_attach.py]: perhaps we should check the file size before attaching it to make sure it is under 19.5MB ?

Comment (alevitter) [reply scripts/bugzilla_attach.py]: Done in recent commit

File: scripts/bugzilla_attach.py (hunk 13 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Done in recent commit

File: scripts/bugzilla_attach.py (hunk 12 above)
Comment (alevitter) [reply scripts/bugzilla_attach.py]: Done in recent commit

File: scripts/bugzilla_attach.py [hunk 15]
```diff
@@ -150,6 +152,9 @@ def main():
     # Now that the archive is created, move the files back in place of the trimmed versions
     restore_hidden_files(logfolder)
 
+    if os.path.getsize(archive_name) > MAX_ARCHIVE_SIZE:
+        print("Archive %s is too large to upload (exceeds %d bytes)." % (archive_name, MAX_ARCHIVE_SIZE))
```
Comment (masayag) [inline scripts/bugzilla_attach.py]: I'd use size in MB instead of bytes
